Rubberduck.Parsing/SelectionExtensions.cs

```diff
@@ -0,0 +1,66 @@
+﻿using Antlr4.Runtime;
+using Rubberduck.VBEditor;
+using System;
+
+namespace Rubberduck.Parsing
+{
+    /// <summary>
+    /// Provide extensions on selections & contexts/tokens
+    /// to assist in validating whether a selection contains
+    /// a given context or a token. 
+    /// </summary>
+    public static class SelectionExtensions
+    {
+        public static bool Contains(this Selection selection, IToken token)
+        {
+            return
+                (((selection.StartLine == token.Line) && (selection.StartColumn - 1) <= token.Column) 
+                    || (selection.StartLine < token.Line))
+             && (((selection.EndLine == token.EndLine()) && (selection.EndColumn - 1) >= (token.EndColumn())) 
+                    || (selection.EndLine > token.EndLine()));
+        }
+
+        public static bool Contains(this ParserRuleContext context, Selection selection)
+        {
+            return
+               (((selection.StartLine == context.Start.Line) && (selection.StartColumn - 1) <= context.Start.Column) 
+                    || (selection.StartLine < context.Start.Line))
+            && (((selection.EndLine == context.Stop.EndLine()) && (selection.EndColumn - 1) >= (context.Stop.EndColumn())) 
+                    || (selection.EndLine > context.Stop.EndLine()));
+        }
+
+        /// <summary>
+        /// Because a token can be spread across multiple lines with line continuations
+        /// it is necessary to do some work to determine the token's actual ending column.
+        /// Whitespace and newline should be preserved within the token.
+        /// </summary>
+        /// <param name="token">The last token within a given context to test</param>
+        /// <returns></returns>
+        public static int EndColumn(this IToken token)
+        {
+            if (token.Text.Contains(Environment.NewLine))
+            {
+                var splitStrings = token.Text.Split(new string[] { Environment.NewLine }, StringSplitOptions.RemoveEmptyEntries);
```

Comment:
I've been thinking about this and I'm worried this could be a bug one day. Can anyone else imagine a valid syntax where a token could be split across newlines and have *empty* newlines (without any line continuation characters)? It is possible to have several blank lines between `If` and `End If` but not so for `End` and `If` in the `End If`.... Or is there?

Reply:
Without a line continuation, splitting `End` and `If` would have the `End` parsed as a keyword, and the `If` parsed as a malformed `If` statement missing a condition and a `Then`.

Reply:
Yes, that's definitely true for the specific case of `End If` - I just want to check if in general could there be any syntax where blanks lines are legal? I think not but want to be 100% sure, given that I'm removing any empty entries in the `Replace()` which could yield different result.

Reply:
The only multi-line token without a line continuations we have in `Rubberduck.Parsing.Grammar.VBALexer.g4` is `NEWLINE`.

Reply:
Should have thought to look there. Good to know I won't need to worry about funky tokens contain empty newlines within its content then.